### PR TITLE
:seedling: Bump to Kubernetes v1.29.1 and controller-runtime v0.17.0

### LINKED ISSUE #4781: Bump k8s to v1.29 and controller-runtime to v0.17.0
/kind feature

In order to to be most useful to our users, we should upgrade Kubernetes to v1.29 and controller-runtime to v0.17.0.

This should likely be in version v2.5.0.

### PR DESCRIPTION
**Pull request by nrb**
**What this PR does / why we need it**:

Updates core dependencies to latest versions.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4781 
**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [x] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)

**Release note**:

```release-note
Support Kubernetes v1.29.x and controller-runtim v0.17.x
```

### CODE REVIEW

**Comment by nrb** [comment]
/hold

Should hold this back for v2.5

**Comment by nrb** [comment]
```
Summarizing 1 Failure:
  [FAIL] [unmanaged] [Cluster API Framework] Clusterctl Upgrade Spec [from latest v1beta1 release to v1beta2] [AfterEach] Should create a management cluster and then upgrade all the providers
  /home/prow/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.6.1/framework/cluster_helpers.go:176 
```

Need to investigate this more closely, but I think this might be related to #4757

File: go.mod [hunk 1]
```diff
@@ -41,18 +41,18 @@ require (
...
+	k8s.io/cli-runtime v0.29.1
+	k8s.io/client-go v0.29.1
+	k8s.io/component-base v0.29.1
+	k8s.io/klog/v2 v2.120.1
 	k8s.io/utils v0.0.0-20240102154912-e7106e64919e
 	sigs.k8s.io/aws-iam-authenticator v0.6.13
 	sigs.k8s.io/cluster-api v1.6.1
 	sigs.k8s.io/cluster-api/test v1.6.1
-	sigs.k8s.io/controller-runtime v0.16.3
+	sigs.k8s.io/controller-runtime v0.17.0
```
Comment (vincepri) [inline go.mod:55]: Cluster API 1.6 is still using 0.16, any specific reason to upgrade this out of band? We tend to match them https://github.com/kubernetes-sigs/cluster-api/blob/release-1.6/go.mod#L49

**Comment by nrb** [comment]
We actually don't need Kube 1.29 til CAPI 1.7. This was mostly to silence Dependabot PRs, but that's been addressed by #4852

**Pull request closed without merging**